invokeai/app/invocations/control_adapter.py

```diff
@@ -0,0 +1,189 @@
+from builtins import float
+from typing import List, Literal, Optional, Union
+
+from pydantic import BaseModel, Field, root_validator, validator
+
+from invokeai.app.invocations.primitives import ImageField
+
+from ...backend.model_management import BaseModelType
+from .baseinvocation import (
+    BaseInvocation,
+    BaseInvocationOutput,
+    FieldDescriptions,
+    Input,
+    InputField,
+    InvocationContext,
+    OutputField,
+    UIType,
+    invocation,
+    invocation_output,
+)
+
+CONTROL_ADAPTER_TYPES = Literal["ControlNet", "IP-Adapter", "T2I-Adapter"]
+
+CONTROLNET_MODE_VALUES = Literal["balanced", "more_prompt", "more_control", "unbalanced"]
+CONTROLNET_RESIZE_VALUES = Literal[
+    "just_resize",
+    "crop_resize",
+    "fill_resize",
+    "just_resize_simple",
+]
+
+
+class ControlNetModelField(BaseModel):
+    """ControlNet model field"""
+
+    model_name: str = Field(description="Name of the ControlNet model")
+    base_model: BaseModelType = Field(description="Base model")
+
+
+class ControlField(BaseModel):
```

Comment:
I don't love that `ControlField` is being used for both ControlNet and IP-Adapter. There are multiple fields that are only relevant to one or the other, which is already confusing and will only get worse as we add T2I-Adapters and more.

I think we should split `ControlField` into separate models for each control adapter type.

---

This also raises the question: Should `DenoiseLatentsInvocation` have a single `control` parameter? Or a separate parameter per control type (e.g. `controlnet_control`, `ip_adapter_control`, etc.)

I like the idea of having separate fields for each, because it makes it simple to define constraints such as:
- Multiple controlnets are allowed, but only one IP-Adapter is allowed.
- Some pairs of control types can be combined while others are mutually exclusive.

I'm definitely open to hearing counter points on this.

Reply:
I have similar concerns, voiced in my review.

I'm in favor of separate fields, with a promise to work out a way for fields to be collapsible in the UI and not end up with physically ginormo nodes.